client_plugin/drivers/shared/shared_driver.go

```diff
@@ -336,43 +336,66 @@ func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	if !d.kvStore.CompareAndPutStateOrBusywait(kvstore.VolPrefixState+r.Name,
 		string(kvstore.VolStateReady),
 		string(kvstore.VolStateDeleting)) {
-		clientFetchSucceeded := true
+		// Failed to change state from Ready to Deleting
+		// 1. Volume is in Mounted state -> get clients and return error
+		// 2. Volume is already in Deleting state and timeout -> continue delete
+		// 3. Volume is in other states -> return error
+
 		// Get a list of host VMs using this volume, if any
 		keys := []string{
+			kvstore.VolPrefixState + r.Name,
 			kvstore.VolPrefixInfo + r.Name,
 		}
 		entries, err := d.kvStore.ReadMetaData(keys)
 		if err != nil {
-			clientFetchSucceeded = false
-			log.Warningf("Failed to check which host VMs are using volume %s", r.Name)
-		}
-		// Unmarshal Info key
-		err = json.Unmarshal([]byte(entries[0].Value), &volRecord)
-		if err != nil {
-			clientFetchSucceeded = false
-			log.Warningf("Failed to unmarshal data. %v", err)
+			msg = fmt.Sprintf("Remove failed: cannot read metadata of volume %s", r.Name)
+			log.Errorf(msg)
+			return volume.Response{Err: msg}
 		}
 
-		msg = fmt.Sprintf("Remove failed: Failed to set volume state to deleting.")
-		if clientFetchSucceeded {
-			msg += fmt.Sprintf(" Containers on other host VM are still using volume %s.",
-				r.Name)
+		state := entries[0].Value
+		log.Infof("State of volume %s is %s", r.Name, state)
+		if state == string(kvstore.VolStateDeleting) {
+			msg = fmt.Sprintf("Volume already in deleting state after timeout. Continue deleting")
+			log.Warningf(msg)
+		} else if state == string(kvstore.VolStateMounted) {
+			// Unmarshal Info key
+			err = json.Unmarshal([]byte(entries[1].Value), &volRecord)
+			if err != nil {
+				msg = fmt.Sprintf("Remove failed: cannot unmarshal info data. %v", err)
+				log.Errorf(msg)
+				return volume.Response{Err: msg}
+			}
+
+			msg = fmt.Sprintf("Remove failed: volume state is Mounted.")
 			msg += fmt.Sprintf(" Host VMs using this volume: %s",
 				strings.Join(volRecord.ClientList, ","))
+			log.Errorf(msg)
+			return volume.Response{Err: msg}
+		} else {
+			msg = fmt.Sprintf("Remove failed: cannot delete from current state.")
+			log.Errorf(msg)
+			return volume.Response{Err: msg}
 		}
-		log.Warningf(msg)
-		return volume.Response{Err: msg}
 	}
 
 	// Delete internal volume
 	log.Infof("Attempting to delete internal volume for %s", r.Name)
 	internalVolname := internalVolumePrefix + r.Name
 	err := d.dockerOps.VolumeRemove(internalVolname)
 	if err != nil {
-		msg = fmt.Sprintf("Failed to remove internal volume %s. Reason: %v", r.Name, err)
-		msg += fmt.Sprintf(" Check the status of the volumes belonging to driver %s", d.internalVolumeDriver)
+		msg = fmt.Sprintf("Failed to remove internal volume from driver %s for volume %s. Reason: %v.",
+			d.internalVolumeDriver, r.Name, err)
+
+		// check the internal volume state:
```

Comment:
When VolumeRemove failed, it could because the volume is in use, or the volume is not accessible.
VolumeInspect can tell us which status the volume is.
If it's still existing (in use), we should stop the deleting process.
Or else, we should proceed to delete the metadata since the internal volume is already inaccessible.